Allow set empty string storage class for legacy

fabrixxm opened the pull request:
Legacy storage is defined by an empty string.
`StorageManager::setBackend()` now accept an empty string as a valid
backend storage.

fabrixxm committed:
Allow set empty string storge class for legacy

Legacy storage is defined by an empty string.
`StorageManager::setBackend()` now accept an empty string as a valid
backend storage.

MrPetovan commented:
Indeed, but this leaves out my initial question: what happens if the legacy storage backend is selected when the migration task is launched?

fabrixxm commented:
> what happens if the legacy storage backend is selected when the migration task is launched?

eeeeh.. mmmh.. uuuh... well..

its *undefined* 

:-P

No, at this moment if you switch back to legacy, data is not moved in database and it keeps to be loaded from the storage backend. The `StorageManager::move()` method will exit throwing an exception [here](https://github.com/friendica/friendica/blob/develop/src/Core/StorageManager.php#L128). 

The correct correct solution should be then support moving back data in database when legacy backend is selected.

MrPetovan commented:
So, we have a legacy value we can't migrate back to and triggers an error when it is saved because it is shown in the admin settings, and multiple additional values that may or may not be selected at any point and more importantly can disappear when the corresponding addon is disabled.

I see couple points of improvement:
- It shouldn't be possible to revert to the legacy value once it has been changed from. This means saving the legacy value is only okay if it was the already existing value.
- Storage backend addons shouldn't be able to be disabled unless the current storage backend setting value has been changed out of its provided value and none of the files are using the addon backend storage anymore.

What do you think?

fabrixxm commented:
> can disappear when the corresponding addon is disabled.

from the select box, but it can still be used by the system (unless you delete the addon)

> It shouldn't be possible to revert to the legacy value once it has been changed from. This means saving the legacy value is only okay if it was the already existing value.

I think there is nothing that stop the move() function to be able to move data back in db, thus allowing users to select Legacy again once they changed it..

> Storage backend addons shouldn't be able to be disabled unless the current storage backend setting value has been changed out of its provided value and none of the files are using the addon backend storage anymore.

I'm ok with this.

MrPetovan commented:
> from the select box, but it can still be used by the system (unless you delete the addon)

How does it work with a disabled backend storage providing addon?

> I think there is nothing that stop the move() function to be able to move data back in db, thus allowing users to select Legacy again once they changed it..

But what would be the point? Isn't the new file system default better? And if it isn't, what's its point?

fabrixxm commented:
> How does it work with a disabled backend storage providing addon?

The db store the full class name of the storage. A disabled backend is removed from the "available bankends" list but when the code loads the class name from the db it just use the class.
I know it's not fully logic (a disabled backend still works ?!?), that's why I'm ok with prevent users to disable addons with storage backends when it's selected as default and there is something in db using it, because it's a nice way to say the user "remember to move your data out of there!".

> But what would be the point? Isn't the new file system default better? And if it isn't, what's its point?

well. yes. true.
so. here a plan:

- in admin don't call `StorageManager::setBackend()` if `$storagebackend` is `""`
- in admin don't show "Legacy" anymore as an option after it has been switched
  - a note should be added about this

this should fix the original issue while we maintain the momentum toward deprecation of legacy storage

MrPetovan commented:
> The db store the full class name of the storage. A disabled backend is removed from the "available backends" list but when the code loads the class name from the db it just use the class.

Ha, this was a point I wanted to reach: how is the addon class autoloaded? Especially if the addon is disabled? The current autoloader only cares about what's in `/src` and a couple other places explicitly mentioned in the base `composer.json` file. I know autoloading feels like magic, but there is a logic to it, and arbitrary classes defined in subfolders of `/addon` wouldn't be picked up by the current autoloading definition.

fabrixxm commented:
I've added a name space in `composer.json` with my initial PR for storage backends : https://github.com/friendica/friendica/commit/8dbedeee5e2dc306e772233dd8207629391c6d9c

MrPetovan commented:
Ooooh.

Nice.

You can scratch off the expected migration on addon disable but we still need to change the setting to a default value because the addon-provided value wouldn't be available anymore since the hook would be removed.

fabrixxm committed:
Revert b2794bb2c 'Allow set empty string storge class for legacy'

fabrixxm committed:
Tweak storage selection in admin page

Don't call `StorageManager::setBackend()` on save when legacy is selected,
dont't show legacy option in selectbox after backend is changed from legacy

fabrixxm committed:
Use direct logic

fabrixxm commented:
I've reverted my first commit and modified admin. I don't have a way to test it atm...

MrPetovan commented on mod/admin.php:
```diff
@@ -1154,40 +1154,41 @@ function update_table(App $a, $table_name, $fields, $old_url, $new_url)
-			switch ($info[0]) { // type
+		if (StorageManager::setBackend($storagebackend)) {
+			$storage_opts = $storagebackend::getOptions();
+			$storage_form_prefix=preg_replace('|[^a-zA-Z0-9]|' ,'', $storagebackend);
```
Code standards: Please wrap operators with spaces.

MrPetovan commented on mod/admin.php:
```diff
@@ -1154,40 +1154,41 @@ function update_table(App $a, $table_name, $fields, $old_url, $new_url)
+			$storage_opts = $storagebackend::getOptions();
+			$storage_form_prefix=preg_replace('|[^a-zA-Z0-9]|' ,'', $storagebackend);
+			$storage_opts_data = [];
+			foreach($storage_opts as $name => $info) {
```
Code standards: Please add a space after `foreach`.

MrPetovan commented on mod/admin.php:
```diff
@@ -1154,40 +1154,41 @@ function update_table(App $a, $table_name, $fields, $old_url, $new_url)
+			foreach($storage_opts as $name => $info) {
+				$fieldname = $storage_form_prefix . '_' . $name;
+				switch ($info[0]) { // type
 				case 'checkbox':
```
Code standard: Please increase the indentation of the switch block.

fabrixxm committed:
Code standards
